File: examples/vslice/src/main/scala/csw/examples/vslice/assembly/TromboneCommandHandler.scala
package csw.examples.vslice.assembly

import akka.actor.{Actor, ActorContext, ActorLogging, ActorRef, ActorSystem, PoisonPill, Props}
import akka.pattern.ask
import akka.util.Timeout
import csw.examples.vslice.assembly.FollowActor.SetZenithAngle
import csw.examples.vslice.assembly.FollowCommand.StopFollowing
import csw.examples.vslice.hcd.TromboneHCD
import csw.examples.vslice.hcd.TromboneHCD._
import csw.services.ccs.CommandStatus._
import csw.services.ccs.SequentialExecutor.{CommandStart, ExecuteOne, StopCurrentCommand}
import csw.services.ccs.MultiStateMatcherActor.StartMatch
import csw.services.ccs.{DemandMatcher, MultiStateMatcherActor, StateMatcher}
import csw.services.ccs.Validation.{RequiredHCDUnavailableIssue, UnsupportedCommandInStateIssue, WrongInternalStateIssue}
import csw.services.events.EventService
import csw.services.loc.LocationService._
import csw.services.loc.LocationSubscriberClient
import csw.util.config.Configurations.SetupConfig
import csw.util.config.StateVariable.DemandState

import scala.concurrent.Await
import scala.concurrent.duration._

/**
 * TMT Source Code: 9/21/16.
 */
class TromboneCommandHandler(ac: AssemblyContext, tromboneHCDIn: Option[ActorRef], allEventPublisher: Option[ActorRef])
    extends Actor with ActorLogging with LocationSubscriberClient with TromboneStateClient {

  import TromboneStateActor._
  import TromboneCommandHandler._
  import ac._
  implicit val system: ActorSystem = context.system

  //override val prefix = ac.info.prefix
  private val badHCDReference = context.system.deadLetters

  private var tromboneHCD: ActorRef = tromboneHCDIn.getOrElse(badHCDReference)
  private def isHCDAvailable: Boolean = tromboneHCD != badHCDReference

  private val badEventService = None
  var eventService: Option[EventService] = badEventService

  // Set the default evaluation for use with the follow command
  private var setElevationItem = naElevation(calculationConfig.defaultInitialElevation)

  // The actor for managing the persistent assembly state as defined in the spec is here, it is passed to each command
  private val tromboneStateActor = context.actorOf(TromboneStateActor.props())

  def receive: Receive = noFollowReceive()

  def handleLocations(location: Location): Unit = {
    location match {

      case l: ResolvedAkkaLocation =>
        log.debug(s"CommandHandler receive an actorRef: ${l.actorRef}")
        tromboneHCD = l.actorRef.getOrElse(badHCDReference)

      case t: ResolvedTcpLocation =>
        log.info(s"Received TCP Location: ${t.connection} from ${sender()}")
        // Verify that it is the event service
        if (t.connection == EventService.eventServiceConnection()) {
          log.info(s"Subscriber received connection: $t from ${sender()}")
          // Setting var here!
          eventService = Some(EventService.get(t.host, t.port))
          log.info(s"Event Service at: $eventService")
        }

      case u: Unresolved =>
        log.info(s"Unresolved: ${u.connection}")
        if (u.connection == EventService.eventServiceConnection()) eventService = badEventService
        if (u.connection.componentId == ac.hcdComponentId) tromboneHCD = badHCDReference

      case default =>
        log.info(s"CommandHandler received some other location: $default")
    }
  }

  def noFollowReceive(): Receive = stateReceive orElse {

    case l: Location =>
      handleLocations(l)

    case ExecuteOne(sc, commandOriginator) =>

      sc.configKey match {
        case ac.initCK =>
          log.info("Init not fully implemented -- only sets state ready!")
          tromboneStateActor ! SetState(cmdItem(cmdReady), moveItem(moveUnindexed), sodiumItem(false), nssItem(false))
          commandOriginator.foreach(_ ! Completed)

        case ac.datumCK =>
          if (isHCDAvailable) {
            log.info(s"Datums State: $currentState")
            val datumActorRef = context.actorOf(DatumCommand.props(sc, tromboneHCD, currentState, Some(tromboneStateActor)))
            context.become(actorExecutingReceive(datumActorRef, commandOriginator))
            self ! CommandStart
          } else hcdNotAvailableResponse(commandOriginator)

        case ac.moveCK =>
          if (isHCDAvailable) {
            val moveActorRef = context.actorOf(MoveCommand.props(ac, sc, tromboneHCD, currentState, Some(tromboneStateActor)))
            context.become(actorExecutingReceive(moveActorRef, commandOriginator))
            self ! CommandStart
          } else hcdNotAvailableResponse(commandOriginator)

        case ac.positionCK =>
          if (isHCDAvailable) {
            val positionActorRef = context.actorOf(PositionCommand.props(ac, sc, tromboneHCD, currentState, Some(tromboneStateActor)))
            context.become(actorExecutingReceive(positionActorRef, commandOriginator))
            self ! CommandStart
          } else hcdNotAvailableResponse(commandOriginator)

        case ac.stopCK =>
          commandOriginator.foreach(_ ! NoLongerValid(WrongInternalStateIssue("Trombone assembly must be executing a command to use stop")))

        case ac.setAngleCK =>
          commandOriginator.foreach(_ ! NoLongerValid(WrongInternalStateIssue("Trombone assembly must be following for setAngle")))

        case ac.setElevationCK =>
          // Setting the elevation state here for a future follow command
          setElevationItem = sc(ac.naElevationKey)
          log.info("Setting elevation to: " + setElevationItem)
          // Note that units have already been verified here
          val setElevationActorRef = context.actorOf(SetElevationCommand.props(ac, sc, tromboneHCD, currentState, Some(tromboneStateActor)))
          context.become(actorExecutingReceive(setElevationActorRef, commandOriginator))
          self ! CommandStart

        case ac.followCK =>
          if (cmd(currentState) == cmdUninitialized
            || (move(currentState) != moveIndexed && move(currentState) != moveMoving)
            || !sodiumLayer(currentState)) {
            commandOriginator.foreach(_ ! NoLongerValid(WrongInternalStateIssue(s"Assembly state of ${cmd(currentState)}/${move(currentState)}/${sodiumLayer(currentState)} does not allow follow")))
          } else {
            // No state set during follow
            // At this point, parameters have been checked so direct access is okay
            val nssItem = sc(ac.nssInUseKey)

            log.info("Set elevation is: " + setElevationItem)

            // The event publisher may be passed in (XXX FIXME? pass in eventService)
            val props = FollowCommand.props(ac, setElevationItem, nssItem, Some(tromboneHCD), allEventPublisher, eventService.get)
            // Follow command runs the trombone when following
            val followCommandActor = context.actorOf(props)
            log.info("Going to followReceive")
            context.become(followReceive(eventService.get, followCommandActor))
            // Note that this is where sodiumLayer is set allowing other commands that require this state
            tromboneStateActor ! SetState(cmdContinuous, moveMoving, sodiumLayer(currentState), nssItem.head)
            commandOriginator.foreach(_ ! Completed)
          }

        case otherCommand =>
          log.error(s"TromboneCommandHandler:noFollowReceive received an unknown command: $otherCommand")
          commandOriginator.foreach(_ ! Invalid(UnsupportedCommandInStateIssue(s"""Trombone assembly does not support the command \"${otherCommand.prefix}\" in the current state.""")))
      }

    case x => log.error(s"TromboneCommandHandler:noFollowReceive received an unknown message: $x")
  }

  def hcdNotAvailableResponse(commandOriginator: Option[ActorRef]): Unit = {
    commandOriginator.foreach(_ ! NoLongerValid(RequiredHCDUnavailableIssue(s"${ac.hcdComponentId} is not available")))
  }

  def followReceive(eventService: EventService, followActor: ActorRef): Receive = stateReceive orElse {
    case ExecuteOne(sc, commandOriginator) =>
      sc.configKey match {
        case ac.datumCK | ac.moveCK | ac.positionCK | ac.followCK | ac.setElevationCK =>
          commandOriginator.foreach(_ ! Invalid(WrongInternalStateIssue("Trombone assembly cannot be following for datum, move, position, setElevation, and follow")))

        case ac.setAngleCK =>
          // Unclear what to really do with state here
          // Everything else is the same
          tromboneStateActor ! SetState(cmdBusy, move(currentState), sodiumLayer(currentState), nss(currentState))

          // At this point, parameters have been checked so direct access is okay
          // Send the SetAngle to the follow actor
          val zenithAngleItem = sc(ac.zenithAngleKey)
          followActor ! SetZenithAngle(zenithAngleItem)
          executeMatch(context, idleMatcher, tromboneHCD, commandOriginator) {
            case Completed =>
              tromboneStateActor ! SetState(cmdContinuous, move(currentState), sodiumLayer(currentState), nss(currentState))
            case Error(message) =>
              log.error(s"setElevation command failed with message: $message")
          }

        case ac.stopCK =>
          // Stop the follower
          log.debug(s"Stop received while following")
          followActor ! StopFollowing
          tromboneStateActor ! SetState(cmdReady, moveIndexed, sodiumLayer(currentState), nss(currentState))

          // Go back to no follow state
          context.become(noFollowReceive())
          commandOriginator.foreach(_ ! Completed)
      }

    case x => log.error(s"TromboneCommandHandler:followReceive received an unknown message: $x")
  }

  def actorExecutingReceive(currentCommand: ActorRef, commandOriginator: Option[ActorRef]): Receive = stateReceive orElse {
    case CommandStart =>
      import context.dispatcher
      implicit val t = Timeout(5.seconds)

      // Execute the command actor asynchronously, pass the command status back, kill the actor and go back to waiting
      for {
        cs <- (currentCommand ? CommandStart).mapTo[CommandStatus]
      } {
        commandOriginator.foreach(_ ! cs)
        currentCommand ! PoisonPill
        self ! CommandDone
      }

    case SetupConfig(ac.stopCK, _) =>
      log.debug("actorExecutingReceive: Stop CK")
      closeDownMotionCommand(currentCommand, commandOriginator)

    case ExecuteOne(SetupConfig(ac.stopCK, _), _) =>
      log.debug("actorExecutingReceive: ExecuteOneStop")
      closeDownMotionCommand(currentCommand, commandOriginator)

    case CommandDone =>
      context.become(noFollowReceive())

    case x => log.error(s"TromboneCommandHandler:actorExecutingReceive received an unknown message: $x")
  }

  private def closeDownMotionCommand(currentCommand: ActorRef, commandOriginator: Option[ActorRef]): Unit = {
    currentCommand ! StopCurrentCommand
    currentCommand ! PoisonPill
    context.become(noFollowReceive())
    commandOriginator.foreach(_ ! Cancelled)
  }

}

object TromboneCommandHandler {

  def props(assemblyContext: AssemblyContext, tromboneHCDIn: Option[ActorRef], allEventPublisher: Option[ActorRef]) =
    Props(new TromboneCommandHandler(assemblyContext, tromboneHCDIn, allEventPublisher))

  def executeMatch(context: ActorContext, stateMatcher: StateMatcher, currentStateSource: ActorRef, replyTo: Option[ActorRef] = None,
                   timeout: Timeout = Timeout(5.seconds))(codeBlock: PartialFunction[CommandStatus, Unit]): Unit = {
    import context.dispatcher
    implicit val t = Timeout(timeout.duration + 1.seconds)

    val matcher = context.actorOf(MultiStateMatcherActor.props(currentStateSource, timeout))
    for {
      cmdStatus <- (matcher ? StartMatch(stateMatcher)).mapTo[CommandStatus]
    } {
      codeBlock(cmdStatus)
      replyTo.foreach(_ ! cmdStatus)
    }
  }

  def idleMatcher: DemandMatcher = DemandMatcher(DemandState(axisStateCK).add(stateKey -> TromboneHCD.AXIS_IDLE))

  def posMatcher(position: Int): DemandMatcher =
    DemandMatcher(DemandState(axisStateCK).madd(stateKey -> TromboneHCD.AXIS_IDLE, positionKey -> position))

  // message sent to self when a command completes
  private case object CommandDone
}

File: examples/vsliceJava/src/main/java/csw/examples/vsliceJava/assembly/TromboneCommandHandler.java
package csw.examples.vsliceJava.assembly;

import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Creator;
import akka.japi.pf.ReceiveBuilder;
import akka.util.Timeout;
import csw.examples.vsliceJava.hcd.TromboneHCD;
import csw.services.ccs.DemandMatcher;
import csw.services.ccs.MultiStateMatcherActor;
import csw.services.ccs.SequentialExecutor.ExecuteOne;
import csw.services.ccs.StateMatcher;
import csw.util.config.BooleanItem;
import csw.util.config.DoubleItem;
import javacsw.services.ccs.JSequentialExecutor;
import javacsw.services.events.IEventService;
import javacsw.services.pkg.ILocationSubscriberClient;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

import static akka.pattern.PatternsCS.ask;
import static csw.examples.vsliceJava.assembly.TromboneStateActor.*;
import static csw.examples.vsliceJava.hcd.TromboneHCD.*;
import static csw.services.ccs.CommandStatus.*;
import static csw.services.ccs.CommandStatus.Error;
import static csw.services.ccs.CommandStatus.Invalid;
import static csw.services.ccs.Validation.*;
import static csw.services.loc.LocationService.*;
import static csw.util.config.Configurations.ConfigKey;
import static csw.util.config.Configurations.SetupConfig;
import static csw.util.config.StateVariable.DemandState;
import static javacsw.services.ccs.JCommandStatus.Cancelled;
import static javacsw.services.ccs.JCommandStatus.Completed;
import static javacsw.util.config.JItems.*;
import static scala.compat.java8.OptionConverters.toJava;

/**
 * TMT Source Code: 9/21/16.
 */
@SuppressWarnings({"OptionalUsedAsFieldOrParameterType", "unused", "WeakerAccess"})
class TromboneCommandHandler extends AbstractActor implements TromboneStateClient, ILocationSubscriberClient {
  private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

  private final AssemblyContext ac;
  private final Optional<ActorRef> allEventPublisher;

  @SuppressWarnings("FieldCanBeLocal")
  private final ActorRef badHCDReference;

  private ActorRef tromboneHCD;

  private boolean isHCDAvailable() {
    return !tromboneHCD.equals(badHCDReference);

  }

  private Optional<IEventService> badEventService = Optional.empty();
  private Optional<IEventService> eventService = badEventService;
//  private boolean isEventServiceAvailable() {
//    return eventService.isPresent();
//  }

  // Set the default evaluation for use with the follow command
  private DoubleItem setElevationItem;

  // The actor for managing the persistent assembly state as defined in the spec is here, it is passed to each command
  private final ActorRef tromboneStateActor;

  @SuppressWarnings("FieldCanBeLocal")
  private TromboneStateActor.TromboneState internalState = TromboneStateActor.defaultTromboneState;

  @Override
  public void setCurrentState(TromboneStateActor.TromboneState ts) {
    internalState = ts;
  }

  private TromboneStateActor.TromboneState currentState() {
    return internalState;
  }

  public TromboneCommandHandler(AssemblyContext ac, Optional<ActorRef> tromboneHCDIn, Optional<ActorRef> allEventPublisher) {
    this.ac = ac;
    badHCDReference = context().system().deadLetters();
    this.tromboneHCD = tromboneHCDIn.orElse(badHCDReference);
    tromboneStateActor = context().actorOf(TromboneStateActor.props());
    this.allEventPublisher = allEventPublisher;
    setElevationItem = AssemblyContext.naElevation(ac.calculationConfig.defaultInitialElevation);
    int moveCnt = 0;
    log.info("System  is: " + context().system());

    subscribeToLocationUpdates();
    context().system().eventStream().subscribe(self(), TromboneState.class);


    receive(noFollowReceive());
  }

  private void handleLocations(Location location) {
    if (location instanceof ResolvedAkkaLocation) {
      ResolvedAkkaLocation l = (ResolvedAkkaLocation) location;
      log.debug("CommandHandler receive an actorRef: " + l.getActorRef());
      tromboneHCD = l.getActorRef().orElse(badHCDReference);

    } else if (location instanceof ResolvedTcpLocation) {
      ResolvedTcpLocation t = (ResolvedTcpLocation) location;
      log.info("Received TCP Location: " + t.connection());
      // Verify that it is the event service
      if (location.connection().equals(IEventService.eventServiceConnection())) {
        log.info("Assembly received ES connection: " + t);
        // Setting var here!
        eventService = Optional.of(IEventService.getEventService(t.host(), t.port(), context().system()));
        log.info("Event Service at: " + eventService);
      }

    } else if (location instanceof Unresolved) {
      log.info("Unresolved: " + location.connection());
      if (location.connection().equals(IEventService.eventServiceConnection()))
        eventService = badEventService;
      if (location.connection().componentId().equals(ac.hcdComponentId))
        tromboneHCD = badHCDReference;

    } else {
      log.info("CommandHandler received some other location: " + location);
    }
  }

  private PartialFunction<Object, BoxedUnit> noFollowReceive() {
    return stateReceive().orElse(ReceiveBuilder.

      match(Location.class, this::handleLocations).

      match(ExecuteOne.class, t -> {

        SetupConfig sc = t.sc();
        Optional<ActorRef> commandOriginator = toJava(t.commandOriginator());
        ConfigKey configKey = sc.configKey();

        if (configKey.equals(ac.initCK)) {
          log.info("Init not fully implemented -- only sets state ready!");
          tromboneStateActor.tell(new SetState(cmdItem(cmdReady), moveItem(moveUnindexed), sodiumItem(false), nssItem(false)), self());
          commandOriginator.ifPresent(actorRef -> actorRef.tell(Completed, self()));

        } else if (configKey.equals(ac.datumCK)) {
          if (isHCDAvailable()) {
            log.info("Datums State: " + currentState());
            ActorRef datumActorRef = context().actorOf(DatumCommand.props(sc, tromboneHCD, currentState(), Optional.of(tromboneStateActor)));
            context().become(actorExecutingReceive(datumActorRef, commandOriginator));
            self().tell(JSequentialExecutor.CommandStart(), self());
          } else hcdNotAvailableResponse(commandOriginator);

        } else if (configKey.equals(ac.moveCK)) {
          if (isHCDAvailable()) {
            ActorRef moveActorRef = context().actorOf(MoveCommand.props(ac, sc, tromboneHCD, currentState(), Optional.of(tromboneStateActor)));
            context().become(actorExecutingReceive(moveActorRef, commandOriginator));
            self().tell(JSequentialExecutor.CommandStart(), self());
          } else hcdNotAvailableResponse(commandOriginator);

        } else if (configKey.equals(ac.positionCK)) {
          if (isHCDAvailable()) {
            ActorRef positionActorRef = context().actorOf(PositionCommand.props(ac, sc, tromboneHCD, currentState(), Optional.of(tromboneStateActor)));
            context().become(actorExecutingReceive(positionActorRef, commandOriginator));
            self().tell(JSequentialExecutor.CommandStart(), self());
          } else hcdNotAvailableResponse(commandOriginator);

        } else if (configKey.equals(ac.stopCK)) {
          commandOriginator.ifPresent(actorRef ->
            actorRef.tell(new NoLongerValid(new WrongInternalStateIssue("Trombone assembly must be executing a command to use stop")), self()));

        } else if (configKey.equals(ac.setAngleCK)) {
          commandOriginator.ifPresent(actorRef ->
            actorRef.tell(new NoLongerValid(new WrongInternalStateIssue("Trombone assembly must be following for setAngle")), self()));

        } else if (configKey.equals(ac.setElevationCK)) {
          // Setting the elevation state here for a future follow command
          setElevationItem = jitem(sc, AssemblyContext.naElevationKey);
          log.info("Setting elevation to: " + setElevationItem);
          // Note that units have already been verified here
          ActorRef setElevationActorRef = context().actorOf(SetElevationCommand.props(ac, sc, tromboneHCD, currentState(), Optional.of(tromboneStateActor)));
          context().become(actorExecutingReceive(setElevationActorRef, commandOriginator));
          self().tell(JSequentialExecutor.CommandStart(), self());

        } else if (configKey.equals(ac.followCK)) {
          if (cmd(currentState()).equals(cmdUninitialized)
            || (!move(currentState()).equals(moveIndexed) && !move(currentState()).equals(moveMoving))
            || !sodiumLayer(currentState())) {
            commandOriginator.ifPresent(actorRef ->
              actorRef.tell(new NoLongerValid(new WrongInternalStateIssue("Assembly state of "
                + cmd(currentState()) + "/" + move(currentState()) + "/" + sodiumLayer(currentState()) + " does not allow follow")), self()));
          } else {
            // No state set during follow
            // At this point, parameters have been checked so direct access is okay
            BooleanItem nssItem = jitem(sc, AssemblyContext.nssInUseKey);

            log.info("Set elevation is: " + setElevationItem);

            // The event publisher may be passed in
            Props props = FollowCommand.props(ac, setElevationItem, nssItem, Optional.of(tromboneHCD), allEventPublisher, eventService.get());
            // Follow command runs the trombone when following
            ActorRef followCommandActor = context().actorOf(props);
            log.info("Going to followReceive");
            context().become(followReceive(followCommandActor));
            // Note that this is where sodiumLayer is set allowing other commands that require this state
//            state(cmdContinuous, moveMoving, sodiumLayer(), jvalue(nssItem));
            tromboneStateActor.tell(new SetState(cmdContinuous, moveMoving, sodiumLayer(currentState()), jvalue(nssItem)), self());
            commandOriginator.ifPresent(actorRef -> actorRef.tell(Completed, self()));
          }

        } else {
          log.error("TromboneCommandHandler2:noFollowReceive received an unknown command: " + t + " from " + sender());
          commandOriginator.ifPresent(actorRef ->
            actorRef.tell(new Invalid(new UnsupportedCommandInStateIssue("Trombone assembly does not support the command " +
              configKey.prefix() + " in the current state.")), self()));
        }
      }).
      matchAny(t ->
        log.warning("TromboneCommandHandler2:noFollowReceive received an unknown message: " + t + " from " + sender())).
      build());
  }

  private void hcdNotAvailableResponse(Optional<ActorRef> commandOriginator) {
    commandOriginator.ifPresent(actorRef -> actorRef.tell(
      new NoLongerValid(new RequiredHCDUnavailableIssue(ac.hcdComponentId.toString() + " is not available")), self()));
  }

  private PartialFunction<Object, BoxedUnit> followReceive(ActorRef followActor) {
    return stateReceive().orElse(ReceiveBuilder.
      match(ExecuteOne.class, t -> {
        SetupConfig sc = t.sc();
        Optional<ActorRef> commandOriginator = toJava(t.commandOriginator());
        ConfigKey configKey = sc.configKey();

        if (configKey.equals(ac.datumCK) || configKey.equals(ac.moveCK) || configKey.equals(ac.positionCK) || configKey.equals(ac.followCK) || configKey.equals(ac.setElevationCK)) {
          commandOriginator.ifPresent(actorRef ->
            actorRef.tell(new Invalid(new WrongInternalStateIssue("Trombone assembly cannot be following for datum, move, position, setElevation, and follow")), self()));

        } else if (configKey.equals(ac.setAngleCK)) {
          // Unclear what to really do with state here
          // Everything else is the same
          tromboneStateActor.tell(new SetState(cmdBusy, move(currentState()), sodiumLayer(currentState()), nss(currentState())), self());

          // At this point, parameters have been checked so direct access is okay
          // Send the SetElevation to the follow actor
          DoubleItem zenithAngleItem = jitem(sc, AssemblyContext.zenithAngleKey);
          followActor.tell(new FollowActor.SetZenithAngle(zenithAngleItem), self());
          Timeout timeout = new Timeout(5, TimeUnit.SECONDS);
          executeMatch(context(), idleMatcher(), tromboneHCD, commandOriginator, timeout, status -> {
            if (status == Completed)
              tromboneStateActor.tell(new SetState(cmdContinuous, move(currentState()), sodiumLayer(currentState()), nss(currentState())), self());
            else if (status instanceof Error)
              log.error("setElevation command failed with message: " + ((Error) status).message());
          });
        } else if (configKey.equals(ac.stopCK)) {
          // Stop the follower
          log.debug("Stop received while following");
          followActor.tell(new FollowCommand.StopFollowing(), self());
          tromboneStateActor.tell(new SetState(cmdReady, moveIndexed, sodiumLayer(currentState()), nss(currentState())), self());

          // Go back to no follow state
          context().become(noFollowReceive());
          commandOriginator.ifPresent(actorRef -> actorRef.tell(Completed, self()));
        }
      }).
      matchAny(t -> log.warning("TromboneCommandHandler:followReceive received an unknown message: " + t)).
      build());
  }

  private PartialFunction<Object, BoxedUnit> actorExecutingReceive(ActorRef currentCommand, Optional<ActorRef> commandOriginator) {
    Timeout timeout = new Timeout(5, TimeUnit.SECONDS);

    return stateReceive().orElse(ReceiveBuilder.
      matchEquals(JSequentialExecutor.CommandStart(), t -> {

        // Execute the command actor asynchronously, pass the command status back, kill the actor and go back to waiting
        ask(currentCommand, JSequentialExecutor.CommandStart(), timeout.duration().toMillis()).
          thenApply(reply -> {
            CommandStatus cs = (CommandStatus) reply;
            commandOriginator.ifPresent(actorRef -> actorRef.tell(cs, self()));
            currentCommand.tell(PoisonPill.getInstance(), self());
            context().become(noFollowReceive());
            return null;
          });
      }).

      match(SetupConfig.class, t -> t.configKey().equals(ac.stopCK), t -> {
        log.debug("actorExecutingReceive: Stop CK");
        closeDownMotionCommand(currentCommand, commandOriginator);
      }).

      match(ExecuteOne.class, t -> {
        log.debug("actorExecutingReceive: ExecuteOneStop");
        closeDownMotionCommand(currentCommand, commandOriginator);
      }).
      matchAny(t -> log.warning("TromboneCommandHandler:actorExecutingReceive received an unknown message: " + t)).
      build());
  }

  private void closeDownMotionCommand(ActorRef currentCommand, Optional<ActorRef> commandOriginator) {
    currentCommand.tell(JSequentialExecutor.StopCurrentCommand(), self());
    currentCommand.tell(PoisonPill.getInstance(), self());
    context().become(noFollowReceive());
    commandOriginator.ifPresent(actorRef -> actorRef.tell(Cancelled, self()));
  }

  // --- static defs ---

  public static Props props(AssemblyContext ac, Optional<ActorRef> tromboneHCDIn, Optional<ActorRef> allEventPublisher) {
    return Props.create(new Creator<TromboneCommandHandler>() {
      private static final long serialVersionUID = 1L;

      @Override
      public TromboneCommandHandler create() throws Exception {
        return new TromboneCommandHandler(ac, tromboneHCDIn, allEventPublisher);
      }
    });
  }

  static void executeMatch(ActorContext context, StateMatcher stateMatcher, ActorRef currentStateSource, Optional<ActorRef> replyTo,
                           Timeout timeout, Consumer<CommandStatus> codeBlock) {

    ActorRef matcher = context.actorOf(MultiStateMatcherActor.props(currentStateSource, timeout));

    ask(matcher, MultiStateMatcherActor.createStartMatch(stateMatcher), timeout).
      thenApply(reply -> {
        CommandStatus cmdStatus = (CommandStatus) reply;
        codeBlock.accept(cmdStatus);
        replyTo.ifPresent(actorRef -> actorRef.tell(cmdStatus, context.self()));
        return null;
      });
  }

  static DemandMatcher idleMatcher() {
    DemandState ds = jadd(new DemandState(axisStateCK.prefix()), jset(stateKey, TromboneHCD.AXIS_IDLE));
    return new DemandMatcher(ds, false);
  }

  static DemandMatcher posMatcher(int position) {
    DemandState ds = jadd(new DemandState(axisStateCK.prefix()),
      jset(stateKey, TromboneHCD.AXIS_IDLE),
      jset(positionKey, position));
    return new DemandMatcher(ds, false);
  }

}

